pd/src/state.rs

```diff
@@ -496,18 +498,27 @@ ON CONFLICT (id) DO UPDATE SET data = $1
     }
 
     /// Retrieve the [`Asset`] for a given asset ID.
-    pub async fn asset_lookup(&self, asset_id: Vec<u8>) -> Result<Asset> {
+    pub async fn asset_lookup(&self, asset_id: Vec<u8>) -> Result<params::AssetInfo> {
         let mut conn = self.pool.acquire().await?;
 
         let asset = query!(
-            "SELECT denom, asset_id FROM assets WHERE asset_id = $1",
+            "SELECT denom, asset_id, total_supply FROM assets WHERE asset_id = $1",
             asset_id
         )
         .fetch_one(&mut conn)
         .await?;
-        Ok(Asset {
-            asset_denom: asset.denom,
-            asset_id: asset.asset_id,
+
+        // TODO: should we be returning proto types from our state methods, or domain types?
+        Ok(params::AssetInfo {
+            denom: Some(
+                asset::REGISTRY
+                    .parse_denom(asset.denom.as_str())
+                    .unwrap()
+                    .into(),
+            ),
+            asset_id: Some(asset::Id::try_from(asset.asset_id)?.into()),
+            total_supply: asset.total_supply.try_into()?, // postgres only has i64....
+            as_of_block_height: 0, // TODO: currently having the caller do this, should we instead pull the latest block height here?
```

Comment:
Yeah, I think that it'd be better to pull the latest block height here, for two reasons: (a) it means that there's only one round-trip to the database, (b) it means that the response data is consistent.

We do all of the writes involved in processing a block transactionally, so although we have no guarantees about what order a given read query will execute relative to some block commit, we do know that each individual query will report consistent state. But if we query the height earlier, then later query the total supply, there could have been an intervening write, causing us to serve inconsistent data.